File: src/ObjectManager/Object.Tes/FilePacks/Records/040-LVLI.Leveled Item.cs
using OA.Core;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class LVLIRecord : Record
    {
        public struct LVLOField
        {
            public short Level;
            public FormId<Record> ItemFormId;
            public int Count;

            public LVLOField(UnityBinaryReader r, int dataSize)
            {
                Level = r.ReadLEInt16();
                r.SkipBytes(2); // Unused
                ItemFormId = new FormId<Record>(r.ReadLEUInt32());
                if (dataSize == 12)
                {
                    Count = r.ReadLEInt16();
                    r.SkipBytes(2); // Unused
                }
                else Count = 0;
            }
        }

        public override string ToString() => $"LVLI: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public BYTEField LVLD; // Chance
        public BYTEField LVLF; // Flags - 0x01 = Calculate from all levels <= player's level, 0x02 = Calculate for each item in count
        public BYTEField? DATA; // Data (optional)
        public List<LVLOField> LVLOs = new List<LVLOField>();

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID": EDID = new STRVField(r, dataSize); return true;
                case "LVLD": LVLD = new BYTEField(r, dataSize); return true;
                case "LVLF": LVLF = new BYTEField(r, dataSize); return true;
                case "DATA": DATA = new BYTEField(r, dataSize); return true;
                case "LVLO": LVLOs.Add(new LVLOField(r, dataSize)); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/040-LVLI.Leveled Item.swift
//
//  LVLIRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class LVLIRecord: Record {
    public struct LVLOField {
        public let level: Int16
        public let itemFormId: FormId<Record>
        public let count: Int16

        init(_ r: BinaryReader, _ dataSize: Int) {
            level = r.readLEInt16()
            r.skipBytes(2) // Unused
            itemFormId = FormId<Record>(r.readLEUInt32())
            if dataSize == 12 {
                count = r.readLEInt16()
                r.skipBytes(2) // Unused
            }
            else { count = 0 }
        }
    }

    public override var description: String { return "LVLI: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var LVLD: BYTEField! // Chance
    public var LVLF: BYTEField! // Flags - 0x01 = Calculate from all levels <= player's level, 0x02 = Calculate for each item in count
    public var DATA: BYTEField? = nil // Data (optional)
    public var LVLOs = [LVLOField]()
    
    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID": EDID = r.readSTRV(dataSize)
        case "LVLD": LVLD = r.readT(dataSize)
        case "LVLF": LVLF = r.readT(dataSize)
        case "DATA": DATA = r.readT(dataSize)
        case "LVLO": LVLOs.append(LVLOField(r, dataSize))
        default: return false
        }
        return true
    }
}
